commons-numbers-complex-arrays/src/main/java/org/apache/commons/numbers/complex/arrays/ComplexList.java

```diff
@@ -0,0 +1,357 @@
+     */
+    private void rangeCheckForInsert(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    /**
+     * Gets the complex number \( (a + i b) \) at the indexed position of the list.
+     *
+     * @param index - Index of the element to get.
+     * @return the complex number.
+     */
+    @Override
+    public Complex get(int index) {
+        rangeCheck(index);
+        final int i = index << 1;
+        return Complex.ofCartesian(realAndImagParts[i], realAndImagParts[i + 1]);
+    }
+
+    /**
+     * Replaces the element at the specified position in this list with the specified element's
+     * real and imaginary parts. No range checks are done.
+     *
+     * @param index - Index of the element to replace.
+     * @param real - Real part \( a \) of the complex number \( (a +ib) \).
+     * @param imaginary - Imaginary part \( b \) of the complex number \( (a +ib) \).
+     */
+    private void setNoRangeCheck(int index, double real, double imaginary) {
+        final int i = index << 1;
+        realAndImagParts[i] = real;
+        realAndImagParts[i + 1] = imaginary;
+    }
+
+    /**
+     * Replaces the element at the specified position in this list with the specified element.
+
+     * @param index - Index of the element to replace.
+     * @param element - Element to be stored at the specified position.
+     * @return the element previously at the specified position.
+     * @throws IndexOutOfBoundsException - if index isn't within the range of list.
+     */
+    @Override
+    public Complex set(int index, Complex element) {
+        rangeCheck(index);
+        final int i = index << 1;
+        final Complex oldValue = Complex.ofCartesian(realAndImagParts[i], realAndImagParts[i + 1]);
+        setNoRangeCheck(index, element.getReal(), element.getImaginary());
+        return oldValue;
+    }
+
+    /**
+     * Increases the capacity of this ComplexList instance, if necessary, to ensure that it can hold at
+     * least the number of elements specified by the minimum capacity argument.
+     *
+     * @param minCapacity – Desired minimum capacity.
+     * @throws OutOfMemoryError - if the {@code minCapacity} is greater than {@code MAX_ARRAY_SIZE}.
+     */
+    public void ensureCapacity(int minCapacity) {
+        ensureCapacityInternal(minCapacity);
+    }
+
+    /**
+     * Increases the capacity of this ComplexList instance, if necessary, to ensure that it can hold at
+     * least the number of elements specified by the minimum capacity argument.
+     *
+     * @param minCapacity – Desired minimum capacity.
+     * @return the backing double array.
+     * @throws OutOfMemoryError - if the {@code minCapacity} is greater than {@code MAX_ARRAY_SIZE}.
+     */
+    private double[] ensureCapacityInternal(int minCapacity) {
+        modCount++;
```

Comment:
This should not be incremented unless the list has been structurally modified, i.e. anything that changes the number of elements. So it should be changed in add/remove methods.

There are cases where this is called and the list will not change number of elements, e.g. ensureCapacity.

You should move the modification of this flag to the appropriate methods.

This behaviour could be tested:

- Create a list
- Get the iterator
- Add an element
- Try and use the iterator.forEach method

It should throw a ConcurrentModificationException